Source/Core/Core/FifoPlayer/FifoRecorder.cpp

```diff
@@ -86,43 +86,77 @@ void FifoRecorder::FifoRecordAnalyzer::OnPrimitiveCommand(OpcodeDecoder::Primiti
   }
   const u32 pos_size = VertexLoader_Position::GetSize(vtx_desc.low.Position, vtx_attr.g0.PosFormat,
                                                       vtx_attr.g0.PosElements);
-  ProcessVertexComponent(CPArray::Position, vtx_desc.low.Position, offset, vertex_size,
-                         num_vertices, vertex_data);
+  const u32 pos_direct_size = VertexLoader_Position::GetSize(
+      VertexComponentFormat::Direct, vtx_attr.g0.PosFormat, vtx_attr.g0.PosElements);
+  ProcessVertexComponent(CPArray::Position, vtx_desc.low.Position, pos_direct_size, offset,
+                         vertex_size, num_vertices, vertex_data);
   offset += pos_size;
 
   const u32 norm_size =
       VertexLoader_Normal::GetSize(vtx_desc.low.Normal, vtx_attr.g0.NormalFormat,
                                    vtx_attr.g0.NormalElements, vtx_attr.g0.NormalIndex3);
-  ProcessVertexComponent(CPArray::Normal, vtx_desc.low.Position, offset, vertex_size, num_vertices,
-                         vertex_data);
+  const u32 norm_direct_size =
+      VertexLoader_Normal::GetSize(VertexComponentFormat::Direct, vtx_attr.g0.NormalFormat,
+                                   vtx_attr.g0.NormalElements, vtx_attr.g0.NormalIndex3);
+  if (vtx_attr.g0.NormalIndex3 && IsIndexed(vtx_desc.low.Normal) &&
+      vtx_attr.g0.NormalElements == NormalComponentCount::NBT)
+  {
+    // We're in 3-index mode, and we're using an indexed format and have the
+    // normal/tangent/binormal, so we actually need to deal with 3-index mode.
+    const u32 index_size = vtx_desc.low.Normal == VertexComponentFormat::Index16 ? 2 : 1;
+    ASSERT(norm_size == index_size * 3);
+    // 3-index mode uses one index each for the normal, tangent and binormal;
+    // the tangent and binormal are internally offset.
+    // The offset is based on the component size, not to the index itself;
+    // for instance, with 32-bit float normals, each normal vector is 3*sizeof(float) = 12 bytes,
+    // so the normal vector is offset by 0 bytes, the tangent by 12, and the binormal by 24.
+    // Using a byte offset instead of increasing the index means that using the same index for all
+    // elements is the same as not using the 3-index mode (increasing the index would give differing
+    // results if the normal array's stride was something other than 12, for instance if vertices
+    // were contiguous in main memory instead of individual components being used).
```

Comment:
Note that I haven't actually hardware-tested how NormalIndex3 works; instead, this is based on how the vertex loaders  implement it: [1](https://github.com/dolphin-emu/dolphin/blob/62601663e578bcd0b981e134c84fbc9d37850a15/Source/Core/VideoCommon/VertexLoader_Normal.cpp#L86) [2](https://github.com/dolphin-emu/dolphin/blob/62601663e578bcd0b981e134c84fbc9d37850a15/Source/Core/VideoCommon/VertexLoaderX64.cpp#L471-L472) [3](https://github.com/dolphin-emu/dolphin/blob/62601663e578bcd0b981e134c84fbc9d37850a15/Source/Core/VideoCommon/VertexLoaderARM64.cpp#L484-L485), which seems to be correct but I haven't actually seen any games where the stride was set up in such a way that it would matter.